File: clients/cpp/src/report.cpp
#include "report.hpp"

#include <algorithm>
#include <string>

namespace {

// Escape a string for embedding in a JSON string literal. Handles the
// characters JSON requires: quote, backslash, and control chars < 0x20. Target
// ids and hex digests never need this, but error reasons (which include curl
// messages) might, so we apply it to all string fields for correctness.
std::string json_escape(const std::string& s) {
    std::string out;
    out.reserve(s.size() + 2);
    for (unsigned char c : s) {
        switch (c) {
            case '"':  out += "\\\""; break;
            case '\\': out += "\\\\"; break;
            case '\b': out += "\\b"; break;
            case '\f': out += "\\f"; break;
            case '\n': out += "\\n"; break;
            case '\r': out += "\\r"; break;
            case '\t': out += "\\t"; break;
            default:
                if (c < 0x20) {
                    static const char* hex = "0123456789abcdef";
                    out += "\\u00";
                    out += hex[c >> 4];
                    out += hex[c & 0x0f];
                } else {
                    out += static_cast<char>(c);
                }
        }
    }
    return out;
}

// Emit one record as a compact JSON object with the contract's fixed key order.
void write_record(std::ostream& out, const Record& r) {
    out << "{\"target\":\"" << json_escape(r.target) << "\""
        << ",\"status\":\"" << r.status << "\""
        << ",\"size_bytes\":" << r.size_bytes
        << ",\"line_count\":" << r.line_count
        << ",\"sha256\":\"" << r.sha256 << "\""
        << ",\"fetch_ms\":" << r.fetch_ms;
    if (r.status != "ok") {
        out << ",\"error\":\"" << json_escape(r.error) << "\"";
    }
    out << "}\n";
}

}  // namespace

void sort_records(std::vector<Record>& records) {
    std::sort(records.begin(), records.end(),
              [](const Record& a, const Record& b) {
                  if (a.size_bytes != b.size_bytes) {
                      return a.size_bytes > b.size_bytes;  // DESC
                  }
                  return a.target < b.target;  // ASC
              });
}

void write_report(std::ostream& out, std::vector<Record>& records) {
    sort_records(records);

    int ok = 0, errors = 0;
    std::int64_t total_bytes = 0;
    for (const auto& r : records) {
        write_record(out, r);
        if (r.status == "ok") {
            ++ok;
            total_bytes += r.size_bytes;
        } else {
            ++errors;
        }
    }

    out << "{\"summary\":true"
        << ",\"total\":" << records.size()
        << ",\"ok\":" << ok
        << ",\"errors\":" << errors
        << ",\"total_bytes\":" << total_bytes
        << "}\n";
}

File: clients/go/report.go
package main

import (
	"bufio"
	"encoding/json"
	"io"
	"sort"
)

// Summary is the footer line emitted after all records. Field order is the JSON
// key order, pinned by the contract (§7.2). Do NOT reorder.
type Summary struct {
	Summary    bool  `json:"summary"`
	Total      int   `json:"total"`
	OK         int   `json:"ok"`
	Errors     int   `json:"errors"`
	TotalBytes int64 `json:"total_bytes"`
}

// sortRecords orders records per §8: size_bytes DESC, then target ASC
// (lexicographic byte order). Error records have size 0 so they naturally sort
// last, ordered among themselves by target ascending. This normalizes the
// nondeterministic completion order of the concurrent workers into a stable,
// identical-across-clients ordering.
func sortRecords(records []Record) {
	sort.Slice(records, func(i, j int) bool {
		if records[i].SizeBytes != records[j].SizeBytes {
			return records[i].SizeBytes > records[j].SizeBytes // DESC
		}
		return records[i].Target < records[j].Target // ASC
	})
}

// writeReport emits the sorted records as JSON Lines followed by the summary
// footer, exactly per the output contract (§7). encoding/json produces compact
// output (no spaces) by default and preserves struct field order, which is what
// the contract requires. Each record is terminated by a single '\n'.
func writeReport(w io.Writer, records []Record) error {
	sortRecords(records)

	bw := bufio.NewWriter(w)
	enc := json.NewEncoder(bw)
	// json.Encoder.Encode writes a trailing newline after each value, which is
	// exactly the JSON Lines framing we want. It does not HTML-escape unless
	// asked; disable escaping so e.g. '<' in a target id (shouldn't happen, but
	// be exact) is not mangled into \u003c.
	enc.SetEscapeHTML(false)

	var sum Summary
	sum.Summary = true
	sum.Total = len(records)
	for _, r := range records {
		if r.Status == "ok" {
			sum.OK++
			sum.TotalBytes += r.SizeBytes
		} else {
			sum.Errors++
		}
		if err := enc.Encode(&r); err != nil {
			return err
		}
	}

	if err := enc.Encode(&sum); err != nil {
		return err
	}

	return bw.Flush()
}
